### PR TITLE
IGNITE-8981: Parallel optimizations for BaggingTrainer

### PR DESCRIPTION
**Pull request by avplatonov**
add Learning environment with Logger and Parallelism strategy while learning

### CODE REVIEW

File: modules/ml/src/main/java/org/apache/ignite/ml/Model.java [hunk 1]
```diff
@@ -32,4 +32,11 @@
     default <X, W> Model<T, X> combine(Model<T, W> other, BiFunction<V, W, X> combiner) {
         return v -> combiner.apply(apply(v), other.apply(v));
     }
+
+    /**
+     * @param pretty Use pretty mode.
+     */
+    default public String toString(boolean pretty) {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/Model.java]: Seems like pretty can be a part of configuration (logging config).

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/BaggingModelTrainer.java [hunk 2]
```diff
@@ -89,11 +91,26 @@ public BaggingModelTrainer(PredictionsAggregator predictionsAggregator,
...
-            learnedModels.add(learnModel(datasetBuilder, featureExtractor, lbExtractor));
-
-        return new ModelsComposition(learnedModels, predictionsAggregator);
+        MLLogger log = environment.logger(getClass());
+        log.log(MLLogger.VerboseLevel.MIN, "Start learning");
+
+        Long startTs = System.currentTimeMillis();
+        List<Promise<ModelOnFeaturesSubspace>> learnedModelsF = new ArrayList<>();
+        for (int i = 0; i < ensembleSize; i++) {
+            learnedModelsF.add(environment.parallelismStrategy().submit(() ->
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/BaggingModelTrainer.java]: Strange part of code, need a time to understand what's happened here

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/ModelsComposition.java [hunk 3]
```diff
@@ -75,4 +76,18 @@ public PredictionsAggregator getPredictionsAggregator() {
...
+        return toString(false);
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString(boolean pretty) {
+        return new StringBuilder("ModelsComposition [").append(pretty ? "\n\t" : "")
+            .append("aggregator = ").append(predictionsAggregator.toString(pretty)).append(pretty ? "\n\t" : ", ")
+            .append("models = [").append(pretty ? "\n\t\t" : "")
+            .append(models.stream().map(x -> x.toString(false)).collect(Collectors.joining(pretty ? "\n\t\t" : ", ")))
+            .append(pretty ? "\n\t]\n]" : "]]").toString();
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/composition/ModelsComposition.java]: Could we say that there are a lot of magic char sequences? Could you move it to constatns or util file?

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/MLLogger.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Helper for ML-specific objects logging.
+ */
+public interface MLLogger {
+    /**
+     * Logging verbose level.
+     */
+    enum VerboseLevel {
+        OFF, MIN, MID, MAX
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/MLLogger.java]: OFF, LOW, HIGH?

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.ml.Model;
+import org.apache.ignite.ml.math.Tracer;
+import org.apache.ignite.ml.math.Vector;
+
+/**
+ * MLLogger implementation based on IgniteLogger.
+ */
+public class OnIgniteMLLogger implements MLLogger {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java]: What the meaning of On here? Could we avoid prepositions?

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * Log line.
+     *
+     * @param verboseLevel Verbose level.
+     * @param line Line.
+     */
+    private void log(VerboseLevel verboseLevel, String line) {
+        switch (verboseLevel) {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java]: Great switch! It helps with real names of verbose levels!

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/parallelism/DefaultParallelismStrategy.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        return new FutureWrapper<>(pool.submit(task::get));
+    }
+
+    /**
+     * Wrapper for future class.
+     *
+     * @param <T>
+     */
+    public static class FutureWrapper<T> implements Promise<T> {
+        private final Future<T> fut;
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/environment/parallelism/DefaultParallelismStrategy.java]: Wow, f is better here :)

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/parallelism/ParallelismStrategy.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+package org.apache.ignite.ml.environment.parallelism;
+
+import org.apache.ignite.ml.math.functions.IgniteSupplier;
+
+/**
+ * Specifies the behaviour of processes in ML-algorithms that can may be parallelized such as parallel learning in
+ * bagging, learning submodels for One-vs-All model, Cross-Validation etc.
+ */
+public interface ParallelismStrategy {
+    public enum Type {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/environment/parallelism/ParallelismStrategy.java]: Maybe we should add here "CUSTOM_POOL" with validation of custom pool checks?

File: modules/ml/src/main/java/org/apache/ignite/ml/tree/DecisionTree.java [hunk 9]
```diff
@@ -250,4 +252,46 @@ private TreeFilter updatePredicateForElseNode(TreeFilter filter, SplitPoint spli
...
+
+    /**
+     * Recursive realisation of DectisionTree to String converting.
+     *
+     * @param node Decision tree.
+     * @param depth Current depth.
+     * @param builder String builder.
+     * @param pretty Use pretty mode.
+     */
+    private static void printTree(DecisionTreeNode node, int depth, StringBuilder builder, boolean pretty) {
```
Comment (zaleslaw) [inline modules/ml/src/main/java/org/apache/ignite/ml/tree/DecisionTree.java]: Make Java code great again! Or Pascal-style

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java (hunk 5 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java]: CustomMLLogger maybe?

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/parallelism/ParallelismStrategy.java (hunk 8 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/environment/parallelism/ParallelismStrategy.java]: It seems to be interesting but CUSTOM_POOL require more information about instantiation it and that's why creation cannot be performed only by enum-value. If user wants custom pool then he/she may implements ParallelismStrategy with custom pool and use .withParallelismStrategy(strategy) method.

File: modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java (hunk 6 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/environment/logging/OnIgniteMLLogger.java]: irony-mode?)

File: modules/ml/src/main/java/org/apache/ignite/ml/composition/BaggingModelTrainer.java (hunk 2 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/composition/BaggingModelTrainer.java]: I have implemented additional default method for Promise that uses list of suppliers
but it doesn't looks better

File: modules/ml/src/main/java/org/apache/ignite/ml/Model.java (hunk 1 above)
Comment (avplatonov) [reply modules/ml/src/main/java/org/apache/ignite/ml/Model.java]: yes I understand
but pretty-parameter affect an internal logic of model visualization that knows nothing about logging
I wanted to implement logic of model representation as unification of ModelFormat but I and @ybabak decided to do this later and implement visualization through ToString-method

**Pull request closed without merging**